### PR TITLE
Stolstov/collection methods

### PR DESCRIPTION
**Pull request by stolstov**
Fixes for:
#176 #177 
#179 

@mbasmanova

### CODE REVIEW

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 1]
```diff
@@ -501,4 +519,303 @@ public int hashCode() {
...
+			
+			if (t == Geometry.Type.Polygon) {
+				if (polygons == null)
+					polygons = new ArrayList<Geometry>();
+				
+				polygons.add(g);
+				continue;
+			}
+			
+			throw new GeometryException("internal error");//what else?
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: perhaps, clarify that unexpected type was encountered

`throw new GeometryException("Internal error: unexpected geometry type: " + t);/`

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 2]
```diff
@@ -501,4 +519,303 @@ public int hashCode() {
...
+	 * an OGCMultiPoint, an OGCMultiPolygon, and an OGCMultilineString.
+	 * @return A flattened Geometry Collection, or self if already flattened.
+	 */
+	public OGCConcreteGeometryCollection flatten() {
+		if (isFlattened()) {
+			return this;
+		}
+		
+		OGCMultiPoint multiPoint = null;
+		ArrayList<Geometry> polygons = null;
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: this can be just `List<Geometry>`

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 3]
```diff
@@ -501,4 +519,303 @@ public int hashCode() {
...
+			}
+
+		}
+
+		return new OGCConcreteGeometryCollection(list, esriSR);
+	}
+	
+	/**
+	 * Fixes topological overlaps in the GeometryCollecion.
+	 * This is equivalent to union of the geometry collection elements.
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: `flatten` method already performs union on polygons; should `flatten` also union points and line strings?

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 4]
```diff
@@ -501,4 +519,303 @@ public int hashCode() {
...
+		for (Geometry g = cursor.next(); g != null; g = cursor.next()) {
+			geoms.add(g);
+		}
+		
+		//make sure geometries don't overlap
+		return removeOverlapsHelper_(geoms);
+	}
+	
+	private OGCConcreteGeometryCollection removeOverlapsHelper_(List<Geometry> geoms) {
+		ArrayList<Geometry> result = new ArrayList<Geometry>();
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: type of `result` could be just `List<Geometry>`

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 5]
```diff
@@ -501,4 +519,303 @@ public int hashCode() {
...
+		for (int i = 0; i < geoms.size(); ++i) {
+			Geometry current = geoms.get(i);
+			if (current.isEmpty())
+				continue;
+			
+			for (int j = i + 1; j < geoms.size(); ++j) {
+				Geometry subG = geoms.get(j);
+				current = OperatorDifference.local().execute(current, subG, esriSR, null);
+				if (current.isEmpty())
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: Why breaking on the first geometry that doesn't overlap `current`? Shouldn't we check for all geometries first?

File: src/main/java/com/esri/core/geometry/OGCStructureInternal.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+		}
+
+		@Override
+		public int getGeometryID() {
+			return m_shape.getGeometryUserIndex(m_geom, m_index);
+		}
+		
+	};
+	
+	public static GeometryCursor prepare_for_ops_(GeometryCursor geoms, SpatialReference sr) {
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/OGCStructureInternal.java]: Looks like this method simplifies the shapes. Would it be clearer to rename it to `simplify` or implement a `SimplifyingGeometryCursor` instead to match `FlatteningGeometryCursor` pattern?

File: src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java [hunk 7]
```diff
@@ -295,10 +295,7 @@ public boolean crosses(OGCGeometry another) {
 	}
 
 	public boolean within(OGCGeometry another) {
-		com.esri.core.geometry.Geometry geom1 = getEsriGeometry();
-		com.esri.core.geometry.Geometry geom2 = another.getEsriGeometry();
-		return com.esri.core.geometry.GeometryEngine.within(geom1, geom2,
-				getEsriSpatialReference());
+		return another.contains(this);
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: `within` is not the same as `contains`, is it? Also, why is it any different than other relationship methods? E.g. Shouldn't `com.esri.core.geometry.ogc.OGCConcreteGeometryCollection` ovewrite it?

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 8]
```diff
@@ -501,4 +519,303 @@ public int hashCode() {
 		
 		return hash;
 	}
+	
+	@Override
+	public double distance(OGCGeometry another) {
+		double minD = 0;
```
[the remark this answers is not in the document]
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: sure

File: src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java [hunk 9]
```diff
@@ -328,6 +325,10 @@ public boolean relate(OGCGeometry another, String matrix) {
 
 	// analysis
 	public double distance(OGCGeometry another) {
+		if (another.geometryType() == OGCConcreteGeometryCollection.TYPE) {
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: Shouldn't the same logic apply to other methods, e.g. `contains`, `overlaps`, etc.?

File: src/main/java/com/esri/core/geometry/ogc/OGCLineString.java [hunk 10]
```diff
@@ -120,7 +121,7 @@ public OGCPoint endPoint() {
 
 	@Override
 	public String geometryType() {
-		return "LineString";
+		return TYPE;
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCLineString.java]: This change is unrelated to the change that adds support for geometry collections. It would be helpful to move it into a separate commit.

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java (hunk 5 above)
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: We break because the current - subG resulted in empty geometry, so current is completely inside of subG.

**Comment by stolstov** [comment]
> I noticed that there are 3 commits, but first commit relies on the new file added in the third commit. Would you merge 1st and 3rd commits?

I'll "squash and merge" this PR.

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 11]
```diff
@@ -955,5 +955,60 @@ public void testEmptyBoundary() throws Exception {
...
+	}
+
+	@Test
+	public void testContainsOnGeometryCollection() {
+		OGCGeometry ogcGeometry = OGCGeometry.fromText("GEOMETRYCOLLECTION (POINT (1 1))");
+		assertTrue(ogcGeometry.contains(OGCGeometry.fromText("POINT (1 1)")));
+	}
+
+	@Test
+	public void testIntersectsOnGeometryCollection() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: I modified the test to ensure that if `intersects` is symmetrical and the test failed with NPE:

```
	@Test
	public void testIntersectsOnGeometryCollection() {
		assertIntersects("GEOMETRYCOLLECTION (POINT (1 1))", "POINT (1 1)", true);
	}

	private void assertIntersects(String wkt, String otherWkt, boolean intersects)
	{
		assertEquals(OGCGeometry.fromText(wkt).intersects(OGCGeometry.fromText(otherWkt)), intersects);
		assertEquals(OGCGeometry.fromText(otherWkt).intersects(OGCGeometry.fromText(wkt)), intersects);
	}
```

Here is the failure:

```
	at com.esri.core.geometry.RelationalOperations.relate(RelationalOperations.java:45)
	at com.esri.core.geometry.OperatorDisjointLocal.execute(OperatorDisjointLocal.java:31)
	at com.esri.core.geometry.GeometryEngine.disjoint(GeometryEngine.java:376)
	at com.esri.core.geometry.ogc.OGCGeometry.disjoint(OGCGeometry.java:275)
	at com.esri.core.geometry.ogc.OGCGeometry.intersects(OGCGeometry.java:280)
	at com.esri.core.geometry.TestOGC.assertIntersects(TestOGC.java:1007)
	at com.esri.core.geometry.TestOGC.testIntersectsOnGeometryCollection(TestOGC.java:1001)
```

**Comment by stolstov** [comment]
@mbasmanova This PR fixes the unit tests that you listed in the issues.

File: src/test/java/com/esri/core/geometry/TestOGC.java (hunk 11 above)
Comment (stolstov) [reply src/test/java/com/esri/core/geometry/TestOGC.java]: ok. I'll fix this.

File: src/main/java/com/esri/core/geometry/ogc/OGCLineString.java (hunk 10 above)
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCLineString.java]: I call geometryType() in the implementation of isFlattened. I've changed the methods to avoid string compare.

File: src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java (hunk 7 above)
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: A.contains(B) == B.within(A), so that's why the change.

Comment (mbasmanova) [reply src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: @stolstov Makes sense. With this change `com.esri.core.geometry.GeometryEngine#within` is no longer used. Should it be removed along with `OperatorWithin` and related classes?

Comment (randallwhitman) [reply src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: Public classes and methods are likely to be used elsewhere.
e.g. https://github.com/Esri/spatial-framework-for-hadoop/blob/1ed3d421eb5c062a5309f6c610cd646210daf889/hive/src/main/java/com/esri/hadoop/hive/ST_Within.java#L20

Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: @mbasmanova Within/Contains, Disjoint/Intersects are a part of the API although they are trivial calls of one another. They are in OGCGeometry because they are required by the OGC spec.

Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: @mbasmanova The primary API for this library is the one in the core.geometry package. The core.geometry.ogc is more like the facade for the use cases that require OGC spec.

Comment (mbasmanova) [reply src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: @stolstov Sergey, thanks for clarifying.

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 12]
```diff
@@ -501,4 +519,303 @@ public int hashCode() {
...
+		public int getGeometryID() {
+			return m_index;
+		}
+		
+	};
+	
+	//Collectively processes group of geometry collections (intersects all segments and clusters points).
+	//Flattens collections, removes overlaps.
+	//Once done, the result collections would work well for topological and relational operations.
+	private List<OGCConcreteGeometryCollection> prepare_for_ops_(List<OGCConcreteGeometryCollection> geoms) {
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: @stolstov This method doesn't seem to be used. The only callers I see are calling `OGCStructureInternal.prepare_for_ops_` directly.

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 13]
```diff
@@ -501,4 +519,303 @@ public int hashCode() {
 		
 		return hash;
 	}
+	
+	@Override
+	public double distance(OGCGeometry another) {
```
[the remark this answers is not in the document]
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: Distance returns NaN for empty input. I'll fix this for collection.

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java (hunk 12 above)
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: I've commited a change to use it. See changes to difference and intersection

File: src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java (hunk 9 above)
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: Yes most need to be changed eventually.

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 14]
```diff
@@ -501,4 +519,303 @@ public int hashCode() {
...
+			OGCGeometry g = geometryN(i);
+			if (g.isEmpty())
+				return false;//no empty allowed
+			
+			String t = g.geometryType();
+			if (t != OGCMultiPoint.TYPE && t != OGCMultiPolygon.TYPE && t != OGCMultiLineString.TYPE)
+				return false;
+		}
+		
+		return true;
```
[the remark this answers is not in the document]
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: Fixed

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: Fixed

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: Fixed

File: src/main/java/com/esri/core/geometry/OGCStructure.java [hunk 15]
```diff
@@ -23,8 +23,13 @@
  */
 package com.esri.core.geometry;
 
+import java.util.ArrayList;
 import java.util.List;
 
+import com.esri.core.geometry.ogc.OGCConcreteGeometryCollection;
```
[the remark this answers is not in the document]
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/OGCStructure.java]: Removed

File: src/test/java/com/esri/core/geometry/TestOGC.java (hunk 11 above)
Comment (stolstov) [reply src/test/java/com/esri/core/geometry/TestOGC.java]: Fixed

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 16]
```diff
@@ -955,5 +955,86 @@ public void testEmptyBoundary() throws Exception {
 		}
 	}
 	
+	@Test
+	public void testUnionPointWithEmptyLineString() {
+		assertUnion("POINT (1 2)", "LINESTRING EMPTY", "GEOMETRYCOLLECTION (POINT (1 2))");
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: I'd expect the union of `POINT (1, 2)` and `LINESTRING EMPTY` to be `POINT (1, 2)`, but this test asserts the result to be `GEOMETRYCOLLECTION (POINT (1 2))`.

Comment (stolstov) [reply src/test/java/com/esri/core/geometry/TestOGC.java]: @mbasmanova Yes, a simpler representation would be better.

**Comment by mbasmanova** [comment]
@stolstov Sergey, are you planning to add more tests? This PR adds a lot of new functionality, but existing tests are spotty. I took a stub at writing tests for the `union` and found that union of a `LINESTRING (0 0, 5 0)` and a `POLYGON ((0 0, 5 0, 5 5, 0 5, 0 0))` returns `GEOMETRYCOLLECTION (LINESTRING (0 0, 5 0), POLYGON ((0 0, 5 0, 5 5, 0 5, 0 0)))` instead of `POLYGON ((0 0, 5 0, 5 5, 0 5, 0 0))`.

Here is the source code for the tests: [TestOGCGeometryCollection.java.txt](https://github.com/Esri/geometry-api-java/files/2063831/TestOGCGeometryCollection.java.txt)

```
assertUnion("LINESTRING (0 0, 5 0)", "POLYGON ((0 0, 5 0, 5 5, 0 5, 0 0))", "POLYGON ((0 0, 5 0, 5 5, 0 5, 0 0))");
```

fails with 

```
org.junit.ComparisonFailure: 
Expected :POLYGON ((0 0, 5 0, 5 5, 0 5, 0 0))
Actual   :GEOMETRYCOLLECTION (LINESTRING (0 0, 5 0), POLYGON ((0 0, 5 0, 5 5, 0 5, 0 0)))
```

**Comment by stolstov** [comment]
@mbasmanova Maria, thanks for the tests. I'll add them and make sure it works.

**Comment by stolstov** [comment]
@mbasmanova Yea, the bug with polygon vs polyline is in the difference operation. Polygon.difference(polyline) does not erase polyline segments on the polygon boundary.

**Comment by stolstov** [comment]
@mbasmanova I've created an issue for the difference: https://github.com/Esri/geometry-api-java/issues/179. I've added the fixes and your test to the PR.

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 17]
```diff
@@ -501,4 +538,445 @@ public int hashCode() {
...
+		OGCConcreteGeometryCollection otherCol = new OGCConcreteGeometryCollection(another, esriSR);
+		OGCConcreteGeometryCollection flattened2 = otherCol.flatten();
+		if (flattened2.isEmpty())
+			return true;
+		
+		for (int i = 0, n2 = flattened2.numGeometries(); i < n2; ++i) {
+			OGCGeometry g2 = flattened2.geometryN(i);
+			boolean good = false;
+			for (int j = 0, n1 = flattened1.numGeometries(); j < n1; ++j) {
+				OGCGeometry g1 = flattened1.geometryN(i);
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: This should be `geometryN(j)` not `..(i)`. 

Also, it may happen that neither `g1` fully contains a `g2`, but collectively a set of `g1`s contain all points in `g2`. Here is a test case:

```
public class TestOGCContains
{
    @Test
    public void testGeometryCollection()
    {
        assertContains("GEOMETRYCOLLECTION (POINT (0 0), LINESTRING (0 1, 5 1))", "GEOMETRYCOLLECTION (MULTIPOINT (0 0, 2 1))");
    }

    private void assertContains(String wkt, String otherWkt)
    {
        OGCGeometry geometry = OGCGeometry.fromText(wkt);
        OGCGeometry otherGeometry = OGCGeometry.fromText(otherWkt);
        assertTrue(geometry.contains(otherGeometry));
        assertTrue(otherGeometry.within(geometry));
    }
}
```

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 18]
```diff
@@ -501,4 +538,445 @@ public int hashCode() {
...
+			return true;
+		OGCConcreteGeometryCollection otherCol = new OGCConcreteGeometryCollection(another, esriSR);
+		OGCConcreteGeometryCollection flattened2 = otherCol.flatten();
+		if (flattened2.isEmpty())
+			return true;
+		
+		for (int i = 0, n1 = flattened1.numGeometries(); i < n1; ++i) {
+			OGCGeometry g1 = flattened1.geometryN(i);
+			for (int j = 0, n2 = flattened2.numGeometries(); j < n2; ++j) {
+				OGCGeometry g2 = flattened2.geometryN(i);
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: This should be geometryN(j) not ..(i).

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 19]
```diff
@@ -501,4 +538,445 @@ public int hashCode() {
...
+			return false;
+		
+		if (this == another)
+			return true;
+
+		//TODO: a simple envelope test
+		
+		OGCConcreteGeometryCollection flattened1 = flatten();
+		if (flattened1.isEmpty())
+			return true;
```
[the remark this answers is not in the document]
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: Yes, thank you

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java (hunk 17 above)
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: @mbasmanova Maria, I've added this test and fixed this. Collection contains now uses another.difference(this).isEmpty().

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java (hunk 18 above)
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: Fixed

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java [hunk 20]
```diff
@@ -501,4 +538,445 @@ public int hashCode() {
...
+		OGCConcreteGeometryCollection otherCol = new OGCConcreteGeometryCollection(another, esriSR);
+		OGCConcreteGeometryCollection flattened2 = otherCol.flatten();
+		if (flattened2.isEmpty())
+			return true;
```
[the remark this answers is not in the document]
Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: Fixed

File: src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java (hunk 18 above)
Comment (mbasmanova) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: @stolstov I'm still seeing `OGCGeometry g2 = flattened2.geometryN(i);` using `i` and not `j`.

Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: @mbasmanova Sorry about that. I've missed this place. Please, change it locally if you plan testing this, I'll fix this later tonight.

Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java]: @mbasmanova Should be fixed now. I've added a test for this case.

File: src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java [hunk 21]
```diff
@@ -295,13 +303,14 @@ public boolean crosses(OGCGeometry another) {
...
-		com.esri.core.geometry.Geometry geom1 = getEsriGeometry();
-		com.esri.core.geometry.Geometry geom2 = another.getEsriGeometry();
-		return com.esri.core.geometry.GeometryEngine.within(geom1, geom2,
-				getEsriSpatialReference());
+		return another.contains(this);
 	}
 
 	public boolean contains(OGCGeometry another) {
+		if (another.geometryType() == OGCConcreteGeometryCollection.TYPE) {
+			return another.contains(this);
```
Comment (mbasmanova) [inline src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: `this.contains(another)` is not the equal to `another.contains(this)`, it is?

The following test in [TestOGCContains.java.txt](https://github.com/Esri/geometry-api-java/files/2079969/TestOGCContains.java.txt) fails:

`assertContains("MULTIPOINT (1 2, 3 4)", "GEOMETRYCOLLECTION (POINT (1 2))");`

Comment (mbasmanova) [reply src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: @stolstov Sergey, thanks. How would you like us to use the fork?

Comment (stolstov) [reply src/main/java/com/esri/core/geometry/ogc/OGCGeometry.java]: @mbasmanova I'd like to use the fork to finish this PR. Once done, I'll merge the forked branch here and then merge this PR. It would also make it simpler for you if you wish to add a commit.

**Comment by mbasmanova** [comment]
@stolstov Sergey, I was busy with https://github.com/prestodb/presto/pull/10731 and somewhat lost track of this PR. What are the next steps?

**Comment by stolstov** [comment]
@mbasmanova I'll merge once you and @randallwhitman are OK it.

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 22]
```diff
@@ -955,5 +954,86 @@ public void testEmptyBoundary() throws Exception {
 		}
 	}
 	
+	@Test
+	public void testUnionPointWithEmptyLineString() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: let's remove this test; it is already included in `TestOGCGeometryCollection#testUnionPoint `

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 23]
```diff
@@ -955,5 +954,86 @@ public void testEmptyBoundary() throws Exception {
...
+		assertUnion("POINT (1 2)", "LINESTRING EMPTY", "POINT (1 2)");
+	}
+
+	@Test
+	public void testUnionPointWithLinestring() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: let's remove this test; it is already included in `TestOGCGeometryCollection#testUnionPoint`

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 24]
```diff
@@ -955,5 +954,86 @@ public void testEmptyBoundary() throws Exception {
...
+		assertUnion("POINT (1 2)", "LINESTRING (3 4, 5 6)", "GEOMETRYCOLLECTION (POINT (1 2), LINESTRING (3 4, 5 6))");
+	}
+
+	@Test
+	public void testUnionLinestringWithEmptyPolygon() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: let's move this to `TestOGCGeometryCollection#testUnionLineString`

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 25]
```diff
@@ -955,5 +954,86 @@ public void testEmptyBoundary() throws Exception {
...
+		assertUnion("MULTILINESTRING ((1 2, 3 4))", "POLYGON EMPTY", "LINESTRING (1 2, 3 4)");
+	}
+
+	@Test
+	public void testUnionLinestringWithPolygon() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: let's move this to `TestOGCGeometryCollection#testUnionLineString`

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 26]
```diff
@@ -955,5 +954,86 @@ public void testEmptyBoundary() throws Exception {
...
+		assertUnion("LINESTRING (1 2, 3 4)", "POLYGON ((0 0, 1 1, 0 1, 0 0))",
+				"GEOMETRYCOLLECTION (LINESTRING (1 2, 3 4), POLYGON ((0 0, 1 1, 0 1, 0 0)))");
+	}
+	
+	@Test
+	public void testUnionGeometryCollectionWithGeometryCollection() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: let's move this to `TestOGCGeometryCollection`

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 27]
```diff
@@ -955,5 +954,86 @@ public void testEmptyBoundary() throws Exception {
...
+		assertEquals(expectedWkt, intersection.asText());
+	}
+	
+	private void assertUnion(String leftWkt, String rightWkt, String expectedWkt) {
+		OGCGeometry union = OGCGeometry.fromText(leftWkt).union(OGCGeometry.fromText(rightWkt));
+		assertEquals(expectedWkt, union.asText());
+	}
+	
+	@Test
+	public void testDisjointOnGeometryCollection() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: let's move this to `TestOGCDisjoint`

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 28]
```diff
@@ -955,5 +954,86 @@ public void testEmptyBoundary() throws Exception {
...
+		OGCGeometry ogcGeometry = OGCGeometry.fromText("GEOMETRYCOLLECTION (POINT (1 1))");
+		assertFalse(ogcGeometry.disjoint(OGCGeometry.fromText("POINT (1 1)")));
+	}
+
+	@Test
+	public void testContainsOnGeometryCollection() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: let's move this to `TestOGCContains`

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 29]
```diff
@@ -955,5 +954,86 @@ public void testEmptyBoundary() throws Exception {
...
+	@Test
+	public void testIntersectsOnGeometryCollection() {
+		OGCGeometry ogcGeometry = OGCGeometry.fromText("GEOMETRYCOLLECTION (POINT (1 1))");
+		assertTrue(ogcGeometry.intersects(OGCGeometry.fromText("POINT (1 1)")));
+		ogcGeometry = OGCGeometry.fromText("POINT (1 1)");
+		assertTrue(ogcGeometry.intersects(OGCGeometry.fromText("GEOMETRYCOLLECTION (POINT (1 1))")));
+	}
+
+	@Test
+	public void testDistanceOnGeometryCollection() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: let's move this to `TestOGCDistance`

File: src/test/java/com/esri/core/geometry/TestOGC.java [hunk 30]
```diff
@@ -955,5 +954,86 @@ public void testEmptyBoundary() throws Exception {
...
+		OGCGeometry ogcGeometry = OGCGeometry.fromText("GEOMETRYCOLLECTION (POINT (1 1))");
+		assertTrue(ogcGeometry.distance(OGCGeometry.fromText("POINT (1 1)")) == 0);
+		
+		//distance to empty is NAN
+		ogcGeometry = OGCGeometry.fromText("GEOMETRYCOLLECTION (POINT (1 1))");
+		assertTrue(Double.isNaN(ogcGeometry.distance(OGCGeometry.fromText("POINT EMPTY"))));
+	}
 	
+	@Test
+	public void testFlattened() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGC.java]: let's move this to `TestOGCGeometryCollectionFlatten`

File: src/test/java/com/esri/core/geometry/TestOGCGeometryCollection.java [hunk 31]
```diff
@@ -0,0 +1,237 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.esri.core.geometry;
+
+import com.esri.core.geometry.ogc.OGCGeometry;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestOGCGeometryCollection {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGCGeometryCollection.java]: let's rename this class to `TestOGCUnion` to match other test classes

File: src/test/java/com/esri/core/geometry/TestOGCGeometryCollection.java [hunk 32]
```diff
@@ -0,0 +1,237 @@
...
+	@Test
+	public void testUnionPoint() {
```
Comment (mbasmanova) [inline src/test/java/com/esri/core/geometry/TestOGCGeometryCollection.java]: let's drop `Union` from method names and rely on the class name `TestOGCUnion` to provide the necessary context
